core/src/subgraph/provider.rs

```diff
@@ -36,11 +35,30 @@ impl<L: LinkResolver> SubgraphProvider<L> {
 
         provider
     }
+
+    fn send_remove_events(
+        &self,
+        name: String,
+        id: String,
+    ) -> impl Future<Item = (), Error = SubgraphProviderError> + Send + 'static {
+        let graphql_remove_sink = self.schema_event_sink.clone();
+        let host_remove_sink = self.event_sink.clone();
+        graphql_remove_sink
+            .send(SchemaEvent::SchemaRemoved(name, id.clone()))
+            .map_err(|e| panic!("Failed to forward schema removal: {}", e))
+            .map(|_| ())
+            .join(
+                host_remove_sink
+                    .send(SubgraphProviderEvent::SubgraphRemoved(id))
+                    .map(|_| ())
+                    .map_err(|e| panic!("Failed to forward subgraph removal: {}", e)),
+            ).map(|_| ())
+    }
 }
 
 impl<L: LinkResolver> SubgraphProviderTrait for SubgraphProvider<L> {
     fn deploy(
-        &self,
+        arc_self: &Arc<Self>,
```

Comment:
IMHO, this should not be a caller concern. I'd rather clone the subgraphs Arc, the resolver and the senders etc. internally than require callers to do `SubgraphProvider::deploy(&provider_arc, ...)`.

Reply:
This is a common pattern in our code and I wanted to try something different here, it's good to discuss whether we like it or not.

I agree call sites are nasty, but we only have one non-test call of this function. I like it because besides cutting 8 lines of `.clone()`s and not having to come up with a good name for each cloned variable, it's easy to break long compositions of futures into methods because `self` can be moved into the closures, as is done here with `send_remove_events`.

There is an RFC for writing `self: Arc<Self>` and another for async/await which would really solve things here but they are far off.

Reply:
On further thought, I could probably get the same benefits by cloning `self` entirely. I'll try that.